File: src/phylo/jax/outside.py
import jax
import jax.numpy as jnp
from ._utils import children_of, token_to_likelihood


def downward_pass(U, logNormU, tree, subMatrices, rootProb, alignment):
    """Compute outside (D) vectors for all nodes.

    D^(n)_a = [sum_j M_{aj}(t_{p,s}) U^(s)_j] * L(x_p|a) *
              [pi_a if p=root else sum_i D^(p)_i M_{ia}(t_{g,p})]

    where s is n's sibling, p is n's parent, g is p's parent,
    and L(x_p|a) is the observation likelihood at the parent node.

    Args:
        U: (*H, R, C, A) inside vectors from upward_pass (rescaled)
        logNormU: (*H, R, C) per-node subtree log-normalizers
        tree: Tree(parentIndex, distanceToParent)
        subMatrices: (*H, R, A, A) substitution probability matrices
        rootProb: (*H, A) equilibrium frequencies
        alignment: (R, C) int32 token alignment

    Returns:
        D: (*H, R, C, A) outside vectors per node (rescaled)
        logNormD: (*H, R, C) per-node outside log-normalizers
    """
    parentIndex = tree.parentIndex
    *H, R, C, A = U.shape

    _, _, sibling = children_of(parentIndex)

    # Observation likelihoods per node: (R, C, A)
    obs_like = token_to_likelihood(alignment, A)

    D = jnp.zeros((*H, R, C, A))
    logNormD = jnp.zeros((*H, R, C))

    # Process nodes 1, 2, ..., R-1 in preorder
    preorder_nodes = jnp.arange(1, R, dtype=jnp.int32)
    parent_of = parentIndex[1:]
    sibling_of = sibling[1:]

    init = (D, logNormD)

    (D, logNormD), _ = jax.lax.scan(
        lambda carry, xs: _downward_step(
            carry, xs, U, logNormU, subMatrices, rootProb, parentIndex, obs_like
        ),
        init,
        (preorder_nodes, parent_of, sibling_of),
    )

    return D, logNormD


def _downward_step(carry, xs, U, logNormU, subMatrices, rootProb, parentIndex, obs_like):
    """Single step of the downward (outside) scan."""
    D, logNormD = carry
    node, parent, sib = xs

    # Sibling contribution: sum_j M_{aj}(t_{p,sib}) * U^(sib)_j
    sib_M = subMatrices[..., sib, :, :]  # (*H, A, A)
    sib_U = U[..., sib, :, :]            # (*H, C, A)
    sib_contrib = jnp.einsum('...ij,...cj->...ci', sib_M, sib_U)  # (*H, C, A)

    # Parent contribution: pi_a if parent is root, else sum_i D^(p)_i M_{ia}(t_{g,p})
    parent_is_root = (parent == 0)

    # Non-root case: sum_i D^(p)_i M_{ia}(t_{g,p})
    parent_M = subMatrices[..., parent, :, :]  # (*H, A, A)
    parent_D = D[..., parent, :, :]            # (*H, C, A)
    prop_down = jnp.einsum('...ci,...ia->...ca', parent_D, parent_M)  # (*H, C, A)

    # Root case: just rootProb
    root_contrib = jnp.broadcast_to(rootProb[..., None, :], prop_down.shape)

    parent_contrib = jnp.where(parent_is_root, root_contrib, prop_down)

    # Include parent's observation likelihood L(x_p|a)
    parent_obs = obs_like[parent]  # (C, A) — broadcast to (*H, C, A)
    parent_contrib = parent_contrib * parent_obs

    # D^(n) = sib_contrib * parent_contrib
    D_raw = sib_contrib * parent_contrib  # (*H, C, A)

    # Compute logNormD for this node
    log_norm_from_sib = logNormU[..., sib, :]         # (*H, C)
    log_norm_from_parent = logNormD[..., parent, :]    # (*H, C)
    log_norm_prior = jnp.where(parent_is_root, 0.0, log_norm_from_parent)
    accumulated = log_norm_from_sib + log_norm_prior

    # Rescale
    maxD = jnp.max(D_raw, axis=-1)
    maxD = jnp.maximum(maxD, 1e-300)
    D_rescaled = D_raw / maxD[..., None]
    log_rescale = jnp.log(maxD)

    logNormD_node = accumulated + log_rescale

    D = D.at[..., node, :, :].set(D_rescaled)
    logNormD = logNormD.at[..., node, :].set(logNormD_node)

    return (D, logNormD), None

File: src/phylo/wasm/src/outside.rs
/// Downward pass (outside algorithm).

use crate::tree::children_of;
use crate::token::token_to_likelihood;

/// Compute outside (D) vectors for all nodes.
///
/// Returns (D, logNormD):
///   D: (R*C*A) flat f64
///   logNormD: (R*C) flat f64
pub fn downward_pass(
    u: &[f64],           // (R*C*A) inside vectors
    log_norm_u: &[f64],  // (R*C)
    parent_index: &[i32],
    sub_matrices: &[f64], // (R*A*A)
    root_prob: &[f64],    // (A,)
    alignment: &[i32],    // (R*C)
    r: usize,
    c: usize,
    a: usize,
) -> (Vec<f64>, Vec<f64>) {
    let (_, _, sibling) = children_of(parent_index);
    let obs_like = token_to_likelihood(alignment, r, c, a);

    let mut d = vec![0.0; r * c * a];
    let mut log_norm_d = vec![0.0; r * c];

    // Preorder: nodes 1..R-1
    for n in 1..r {
        let p = parent_index[n] as usize;
        let sib = sibling[n] as usize;
        let sib_m_base = sib * a * a;
        let par_m_base = p * a * a;

        for col in 0..c {
            let sib_u_base = (sib * c + col) * a;
            let par_d_base = (p * c + col) * a;
            let par_obs_base = (p * c + col) * a;

            let mut d_raw = vec![0.0; a];
            for aa in 0..a {
                // Sibling contribution
                let mut sib_contrib = 0.0;
                for j in 0..a {
                    sib_contrib += sub_matrices[sib_m_base + aa * a + j] * u[sib_u_base + j];
                }

                // Parent contribution
                let parent_contrib;
                if p == 0 {
                    parent_contrib = root_prob[aa];
                } else {
                    let mut pc = 0.0;
                    for i in 0..a {
                        pc += d[par_d_base + i] * sub_matrices[par_m_base + i * a + aa];
                    }
                    parent_contrib = pc;
                }

                d_raw[aa] = sib_contrib * parent_contrib * obs_like[par_obs_base + aa];
            }

            // logNormD
            let log_norm_sib = log_norm_u[sib * c + col];
            let log_norm_prior = if p == 0 { 0.0 } else { log_norm_d[p * c + col] };
            let accumulated = log_norm_sib + log_norm_prior;

            // Rescale
            let mut max_val = d_raw[0];
            for aa in 1..a {
                if d_raw[aa] > max_val { max_val = d_raw[aa]; }
            }
            if max_val < 1e-300 { max_val = 1e-300; }

            let node_d_base = (n * c + col) * a;
            for aa in 0..a {
                d[node_d_base + aa] = d_raw[aa] / max_val;
            }
            log_norm_d[n * c + col] = accumulated + max_val.ln();
        }
    }

    (d, log_norm_d)
}
